doc: I2P documentation updates

jonatack opened the pull request:
Address the documentation updates requested in issue #26754, clarify/simplify the -i2pacceptincoming help, and a few other fixups.

jonatack committed:
doc: update/clarify/de-emphasize I2P transient address section

jonatack committed:
doc: update bandwidth section of I2P documentation

jonatack commented on doc/i2p.md:
```diff
@@ -100,7 +79,35 @@ In general, a node can be run with both onion and I2P hidden services (or
+## Persistent vs transient I2P addresses
+
+The first time Bitcoin Core connects to the I2P router, it automatically
+generates a persistent I2P address and its corresponding private key by default
```
Note to self: re-verify if the `-i2pacceptincoming` help ought to be updated regarding "If this is not set but -i2psam is set then only outgoing connections will be made to the I2P network". IIRC listening is the default behavior.

1440000bytes replied:
Yes it should be updated as its confusing at this moment. Incoming i2p connections work by default if i2psam is set with i2p working correctly. Maybe we could change this to:

```
     Incoming I2P connections are accepted via the SAM proxy
     by default. If this is set to 0 and -i2psam is set
     then only outgoing connections will be made to the I2P network.
     Ignored if -i2psam is not set. Listening for incoming I2P
     connections is done through the SAM proxy, not by binding to a
     local address and port (default: 1)
     
```

vasild replied:
The text is correct, but I can see how it can be confusing. By "to be set" I meant "to be enabled" or "=1" and "not set" => "disabled" or "=0". Thus the current text:

> If **set** and **-i2psam is also set** then incoming I2P connections are accepted via the SAM proxy. If **this is not set** but **-i2psam is set** then only outgoing connections will be made to the I2P network.

means:

> If **-i2pacceptincomng=1** and **-i2psam=1** then incoming I2P connections are accepted via the SAM proxy. If **-i2pacceptincoming=0** but **-i2psam=1** then only outgoing connections will be made to the I2P network.

I think the text by @1440000bytes is better than the current text. Maybe it can be improved further - to talk about the "default" only at the end. Like "if 5, then it is blue, if 6 then red and if 7 - green. Default 6." rather than "by default it is red, if 5 - blue, if 7 - green. Default 6"

zzzi2p commented:
Thanks for this. You're on the right track here and we also appreciate the updates on your blog.

Not sure of the audience (users vs. downstreams) for this doc vs. your blog,
or whether this doc is more about your current release or your next one.
So I won't offer a detailed review.

Elaborating my request from OP of the ticket, overall, the messaging that would be most helpful is:

- Emphasize that I2P is a P2P network (just like Bitcoin!), encourage high share limits (or discourage limit setting at all) to protect anonymity and contribute to the network
- Be sure to run latest version of Java I2P (2.1.0 or later) or i2pd (2.45.0 or later)
- For now, discourage -i2pacceptincoming=0, especially on low-performance hardware or low-bandwidth, until we work through the issues.
- I don't know if anybody is trying to bootstrap the whole blockchain over i2p, or if that's even possible, but that's a really bad idea also

Thanks again.

1440000bytes commented:
> whether this doc is more about your current release or your next one

Its the only i2p doc in this repository that users and developers can refer, update etc. for minimal and important things required to use i2p with bitcoin core. Not release specific. It was added in https://github.com/bitcoin/bitcoin/pull/22250 (2021) and updated multiple times since then.

There is also a doc for Tor: [/doc/tor.md](https://github.com/bitcoin/bitcoin/blob/master/doc/tor.md)

jonatack committed:
doc: remove recommended I2P router versions

as these go stale and users will generally install the current versions available.

vasild commented on doc/i2p.md:
```diff
@@ -100,7 +79,35 @@ In general, a node can be run with both onion and I2P hidden services (or
+address is used for making outgoing connections and accepting incoming
+connections.
+
+In the I2P network, a the receiver of an inbound connection sees the address of
```
`s/a the receiver/the receiver/`

jonatack replied:
Thanks, done.

vasild reviewed (approved):
ACK 92f43b479a9673f5475d0de9c731b9f7a37a1c86 modulo the typo mentioned below

jonatack committed:
doc: clarify -i2pacceptincoming help documentation

and also hoist the default setting to a constexpr and
remove unused f-string operators in a related functional test.

jonatack commented:
>     * Emphasize that I2P is a P2P network (just like Bitcoin!), encourage high share limits (or discourage limit setting at all) to protect anonymity and contribute to the network

Done.
 
>     * Be sure to run latest version of Java I2P (2.1.0 or later) or i2pd (2.45.0 or later)

Dropped the versions in the documentation as they go stale. Presumably users will install current versions.
 
>     * For now, discourage -i2pacceptincoming=0, especially on low-performance hardware or low-bandwidth, until we work through the issues.

Done.

>     * I don't know if anybody is trying to bootstrap the whole blockchain over i2p, or if that's even possible, but that's a really bad idea also

Indeed, this was discouraged in the documentation, as it takes a long time.

jonatack commented:
Thanks for the feedback, everyone. Pushed an update and brought this out of draft.

1440000bytes reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26838/commits/3e1d2941e921f96c489c5101e2a7a7bf7e7d8b73

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26838/commits/3e1d2941e921f96c489c5101e2a7a7bf7e7d8b73

These changes improve the I2P documentation.

w0xlt commented:
>  don't know if anybody is trying to bootstrap the whole blockchain over i2p, or if that's even possible, but that's a really bad idea also

Maybe this shouldn't be allowed by code ?

jonatack commented:
> Maybe this shouldn't be allowed by code ?

I thought about this but would prefer not to limit it unless it's been shown that this is really a causing a serious issue with the network, both in principle to not make an exception for one network, and also doing it several releases after adding I2P support, as the very few (if any) users that do would see their nodes not working anymore after upgrading Bitcoin Core.

jonatack commented:
> Maybe this shouldn't be allowed by code ?

I thought about this but would prefer not to limit it unless it's been shown that this is really causing a serious issue with the network, both in principle to not make an exception for one network, and also doing it several releases after adding I2P support, as any users that do would see their nodes not working anymore after upgrading to that release of Bitcoin Core.

vasild reviewed (approved):
ACK 3e1d2941e921f96c489c5101e2a7a7bf7e7d8b73

zzzi2p commented:
ACK

willcl-ark commented:
ACK 3e1d294

Iin the last 24h of running my node I see the following stats via nethogs, for Bitcoin Core v24.0 with ~11 I2P peers, ~85 HTTP peers and ~13 onion peers:

![Selection_048](https://user-images.githubusercontent.com/6606587/211798016-d405a77c-6c8c-467c-97b5-d5c5d8ca43a8.png)

For contrast Tor has used ~600 MB of data total (figure shared with core-lightning). This is roughly the same usage pattern I saw which led me to try and limit I2P traffic only to bitcoin peers initially -- in fact a few months ago when monitoring for longer periods, I found I2P using more than double the bitcoind totals.

I accept (and agree!) that we should want to and encourage contributing back to the wider I2P network for it's network health and, for me personally it's no issue, as I'm on an unmetered connection. But I still wonder whether a comment around "expected" I2P bandwidth usage might be warranted in a future PR to give a heads up to entities with limited resources?

jonatack commented:
@willcl-ark Interesting feedback, worth measuring how much bandwidth is used before/after optimizations like #26837.

jonatack commented:
@willcl-ark Interesting feedback (thanks!), worth measuring how much bandwidth is used after  https://github.com/bitcoin/bitcoin/pull/26837 and upgrading to the latest releases, currently Java I2P 2.1.0 and i2pd 2.45.1.
